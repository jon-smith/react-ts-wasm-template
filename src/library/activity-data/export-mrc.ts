import { Interval } from './interval';

type CourseData = {
	timeMinutes: number;
	intensityPercent: number;
};

const calculateWithStartEndTimes = (intervals: Interval[]) => {
	const withStartEndTimes = intervals.map((i) => ({
		...i,
		startTimeMinutes: 0,
		endTimeMinutes: 0,
	}));
	for (let i = 0; i < intervals.length; ++i) {
		withStartEndTimes[i].startTimeMinutes = i === 0 ? 0 : withStartEndTimes[i - 1].endTimeMinutes;
		withStartEndTimes[i].endTimeMinutes =
			withStartEndTimes[i].startTimeMinutes + withStartEndTimes[i].durationSeconds / 60;
	}
	return withStartEndTimes;
};

const convertToCourseData = (intervals: Interval[]): CourseData[] => {
	const withStartAndEndTimes = calculateWithStartEndTimes(intervals);

	return withStartAndEndTimes.flatMap((t) => [
		{ timeMinutes: t.startTimeMinutes, intensityPercent: t.intensityPercent },
		{ timeMinutes: t.endTimeMinutes, intensityPercent: t.intensityPercent },
	]);
};

const makeCourseHeader = (name: string, description: string) =>
	`
[COURSE HEADER]\n
VERSION = 2\n
UNITS = ENGLISH\n
DESCRIPTION = ${description}\n
FILE NAME = ${name}\n
MINUTES PERCENT\n
[END COURSE HEADER]\n
`;

const makeMRCString = (header: string, data: string) =>
	`
${header}
[COURSE DATA]\n
${data}
[END COURSE DATA]
`;

export const buildMRCFileString = (name: string, description: string, intervals: Interval[]) => {
	const courseHeader = makeCourseHeader(name, description);

	const courseData = convertToCourseData(intervals);
	const courseDataString = courseData
		.map((d) => `${d.timeMinutes.toFixed(2)}\t${d.intensityPercent.toFixed(0)}\n`)
		.join('');

	return makeMRCString(courseHeader, courseDataString);
};
