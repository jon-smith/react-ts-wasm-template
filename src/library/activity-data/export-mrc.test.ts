import { buildMRCFileString } from './export-mrc';

describe('MRC Export', () => {
	it('Should match snapshot', () => {
		const intervals = [
			{ durationSeconds: 60, intensityPercent: 100 },
			{ durationSeconds: 30, intensityPercent: 120 },
			{ durationSeconds: 40, intensityPercent: 80 },
		];
		expect(buildMRCFileString('name', 'description', intervals)).toMatchSnapshot();
	});
});
