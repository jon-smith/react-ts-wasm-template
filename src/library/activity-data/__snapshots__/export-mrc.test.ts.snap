// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MRC Export Should match snapshot 1`] = `
"

[COURSE HEADER]

VERSION = 2

UNITS = ENGLISH

DESCRIPTION = description

FILE NAME = name

MINUTES PERCENT

[END COURSE HEADER]


[COURSE DATA]

0.00	100
1.00	100
1.00	120
1.50	120
1.50	80
2.17	80

[END COURSE DATA]
"
`;
